tools/gmhazard_scripts/db_creation/empirical_db/common.py

```diff
@@ -241,71 +313,30 @@ def calculate_emp_site(
-        )
-
-        site.Rjb = row["rjb"]
-        site.Rrup = row["rrup"]
-        # Currently 0 as Rtvz calculation is done in the empirical engine
-        site.Rtvz = 0  # row["rtvz"]
-        site.fpeak = np.array([12])
-        rx = row["rx"]
-        site.Rx = 0 if np.isnan(rx) else rx
-        ry = row["ry"]
-        site.Ry = 0 if np.isnan(ry) else ry
-
-        rupture_id = rupture_df[
-            rupture_df["rupture_name"] == rupture_name
-        ].index.values.item()
-
-        fault_im_result_dict = {key: {} for key in imdb_dict.keys()}
-
-        for im_type in im_types:
-            GMMs = empirical_factory.determine_all_gmm(
-                fault, str(im_type), tect_type_model_dict
-            )
-            for GMM, __comp in GMMs:
-                db_type = f"{GMM.name}_{fault.tect_type.name}"
-                values = empirical_factory.compute_gmm(
-                    fault,
+    with mp.Pool(n_procs) as p:
+        results = p.starmap(
+            __process_rupture,
+            [
+                (
+                    rupture,
                     site,
-                    GMM,
-                    str(im_type),
-                    psa_periods if im_type is IMType.pSA else None,
+                    rupture_df,
+                    im_types,
+                    tect_type_model_dict,
+                    psa_periods,
+                    keep_sigma_components,
+                    {key: {} for key in imdb_dict.keys()},
                 )
-                if not im_type is IMType.pSA:
-                    values = [values]
-                for i, value in enumerate(values):
-                    full_im_name = (
-                        IM(im_type, period=psa_periods[i])
-                        if im_type is IMType.pSA
-                        else IM(im_type)
-                    )
-                    mean = np.log(value[0])
-                    stdev, sigma_inter, sigma_intra = value[1]
-
-                    fault_im_result_dict[db_type][str(full_im_name)] = mean
-                    if keep_sigma_components:
-                        fault_im_result_dict[db_type][
-                            f"{full_im_name}_sigma_inter"
-                        ] = sigma_inter
-                        fault_im_result_dict[db_type][
-                            f"{full_im_name}_sigma_intra"
-                        ] = sigma_intra
-                    else:
-                        fault_im_result_dict[db_type][f"{full_im_name}_sigma"] = stdev
-        for imdb_key in imdb_dict.keys():
-            im_result_dict[imdb_key][rupture_id] = fault_im_result_dict[imdb_key]
+                for index, rupture in matching_df.iterrows()
+            ],
+        )
 
+    im_result_dict = {key: {} for key in imdb_dict.keys()}
+    for cur_rupture_id, cur_fault_im_dict in results:
+        for cur_gmm, cur_im_dict in cur_fault_im_dict.items():
+            im_result_dict[cur_gmm][cur_rupture_id] = cur_im_dict
```

Comment:
I remember the below df.from_dict being slow too. So if that is out of the mp pool there will be some waiting while that gets done. 

Short of creating a fixed sized df beforehand, I can't think of anything.

Reply:
Meh it might slow, but is only done a couple of times. There might be a faster way, but I don't think its worth it.
From memory, when I ran it with MP the slowest part by far is still the loop over ruptures (which now uses MP)

Reply:
Oh yeah. It's only run once per station. Which isn't much for your use case.